### ISSUE TITLE
Mark HasDataLoaderMeta type as unique

### ISSUE THREAD

**Issue by minht11** [posva/unplugin-vue-router#133]
Currently trying to use `` as interface type gives following error `computed property name in an interface must refer to an expression whose type is a literal type or a 'unique symbol' type.`
Use case trying to declare global meta params
```ts
declare module 'vue-router' {
	interface RouteMeta {
		[HasDataLoaderMeta]: (() => Promise<unknown>)[]
	}
}
```

https://github.com/posva/unplugin-vue-router/blob/f268adbf3a84e3e86aad78f2a039017af17d5276/src/data-fetching/dataFetchingGuard.ts#L6
Can be fixed with
```ts
export const HasDataLoaderMeta: unique symbol = Symbol()
```

**Comment by posva** [comment]
What are you trying to do? This shouldn't be necessary

**Comment by minht11** [comment]
I am using [typesafe-i18n](https://github.com/ivanhofer/typesafe-i18n). I kinda repurposed defineLoader to load namespaced translations async during page load in order to reduce waterfalls.
My code looks basically like this:

```ts

declare module 'vue-router' {
	// TODO. Until is fixed https://github.com/posva/unplugin-vue-router/issues/133
	// eslint-disable-next-line @typescript-eslint/no-shadow
	const HasDataLoaderMeta: unique symbol

	interface RouteMeta {
		[HasDataLoaderMeta]?: (() => Promise<unknown>)[]
		i18nNamespace?: string
	}
}

const useI18nLoader = defineLoader(async (route) => {
	const namespace = route.meta.i18nNamespace as Namespaces

	if (namespaces.includes(namespace)) {
		const { commitLoadedLocales, locale } = useGlobalStore()

		await loadNamespaceAsync(locale, namespace)
		commitLoadedLocales()
	}
})

const i18nLoaderModulePromise = Promise.resolve({
	useI18nLoader,
})

const i18nLoaderModuleFn = () => i18nLoaderModulePromise

export const router = createRouter({
	extendRoutes(routes) {
		// Only iterating over top routes is enough,
		// because nested routes inherit meta from parent.
		for (const route of routes) {
			route.meta ??= {}
			route.meta[HasDataLoaderMeta] ??= []

			route.meta[HasDataLoaderMeta].push(i18nLoaderModuleFn)
		}

		return routes
	},
	history: createWebHistory(),
})
```

**Comment by posva** [comment]
I see. I don't recommend hijacking the RouteMeta with your own type and import as this is considered internal API and if it changes in the future it would break your code. Instead, create another symbol and add another navigation guard.

But even though, `HasDataLoaderMeta` is already a unique symbol as it's created with `= Symbol()`. Otherwise it would error in this project too

**Comment by minht11** [comment]
Typescript doesn't consider it unique, thats why you need to mark it as such. At least not as key for interfaces.

I want to leverage loader so data/locales fetching happens in parallel, so I need a way to add them to loader. If I add separate symbol and load translations myself, I lose pralelization benefits. Loading becomes blocking. I looked into RFC and how to manually add loaders. If that is not recomended perpahs another api is needed?

**Comment by posva** [comment]
> Typescript doesn't consider it unique, thats why you need to mark it as such. At least not as key for interfaces.

It seems to be doing so within the project as `= Symbol()` already forges a `unique symbol` type, so I'm surprised that changes fixes your issue as it doesn't change the type and that makes sense as the project itself would otherwise also error.

Regarding your use case: I don’t think parallelism brings much but you could still await within a loader, for example:

```ts
const useI18nLoader = defineLoader(async (route) => {
	const namespace = route.meta.i18nNamespace as Namespaces
	if (namespaces.includes(namespace)) {
		const { commitLoadedLocales, locale } = useGlobalStore()
        await i18nLoaderModuleFn()
		await loadNamespaceAsync(locale, namespace)
		commitLoadedLocales()
	}
})
```

This would avoid hijacking a private API while keeping the parallelism, so It's future proof 💯

**Comment by minht11** [comment]
You are absolutely right about symbol, sorry. It is another issue entirely.`typed-router.d.ts` inside module declarations exports things from `unplugin-vue-router/runtime`, but in my project those types are not being resolved. If I manually add `import 'unplugin-vue-router/runtime` at the top of `typed-router.d.ts` thise exports start to work. I am not sure how that supposed to behave.

Regarding with i18n loader how do I make loader run on every route? I already await things that i need inside the loader, and  `i18nLoaderModuleFn` is just fake module mock that I later inject to `HasDataLoaderMeta` array , calling it inside the loader itself doesn't make much sense for me or am I misunderstanding something?

**Comment by posva** [comment]
Regarding your loader for each route, the way you are doing it is actually fine. I was confused because you overrode the RouteMeta interface but I also realize the type is probably not augmented correctly.

**Comment by minht11** [comment]
I created [reproduction](https://stackblitz.com/edit/vitejs-vite-crc1ef?file=src%2Fmain.ts,typed-router.d.ts&terminal=dev) of the issue. Stackblitz doesn't seem to support type info in vue files, but nonentless, if you go to the `main.ts` file and hover on meta field you will see nothing. However if you add `import 'unplugin-vue-router/runtime'` at the top of `typed-router.d.ts` it will begin to work.

<img width="221" alt="image" src="https://user-images.githubusercontent.com/8914032/231152181-20917e13-3ebd-4a54-8716-d012ca2ba77e.png">
<img width="370" alt="image" src="https://user-images.githubusercontent.com/8914032/231152513-f502a0f3-ec4c-43d1-b8ce-ecd05ca163ba.png">

**Comment by lioneaglesolutions** [comment]
Not sure if this is the correct issue but is it necessary to have to import the macro from 'vue-router/auto'? Shouldn't it just be available similar to `defineProps`  in terms of typescript?

**Comment by posva** [comment]
Yes, it’s necessary. It’s different from defineProps because it’s a Vue compiler macro

**Comment by posva** [comment]
Heads up: the internal `HasDataLoaderMeta` will be removed in the next releases as it won't be necessary anymore
